### DISCUSSION TITLE
Consitency of API signatures

### DISCUSSION THREAD

**Discussion by afpineda** [h2zero/NimBLE-Arduino#807] [General]
Hi. First things first:
- I apologise for opening a new, rather trivial, question. You have too many of them.
- Congratulations on the release of version 2.1.0 and thanks a lot for your hard work.

In my opinion, version 2.1.0 still has some API inconsistencies that I don't understand. Some signatures seem to be C-oriented, but others seem to be C++-oriented. For example, let's look at this [callback](https://h2zero.github.io/NimBLE-Arduino/class_nim_b_l_e_characteristic_callbacks.html):

```c++
virtual void onWrite (NimBLECharacteristic *pCharacteristic, NimBLEConnInfo &connInfo)
```

It's obvious that `pCharacteristic` can't be null in this callback, but the C-oriented declaration suggests that you should check for null pointers first. The second argument, `connInfo` is just the opposite. It can't be null.

So, to be consistent, the signature should look like this:

```c++
virtual void onWrite (NimBLECharacteristic &pCharacteristic, NimBLEConnInfo &connInfo)
```
or
```c++
virtual void onWrite (const NimBLECharacteristic &pCharacteristic, NimBLEConnInfo &connInfo)
```

This makes it clear which arguments can be null and which cannot. In fact, null pointers are not allowed by the compiler in this way. You have one less check and the runtime error becomes a compile-time error.

What do you think? Am I wrong?

**Comment by h2zero** [comment]
Yes you are correct and I missed this for sure. With all the other changes I was focused on this got overlooked.
